streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java

```diff
@@ -53,7 +54,7 @@
 import static org.apache.kafka.common.utils.Utils.getPort;
 import static org.apache.kafka.streams.processor.internals.InternalTopicManager.WINDOW_CHANGE_LOG_ADDITIONAL_RETENTION_DEFAULT;
 
-public class StreamPartitionAssignor implements PartitionAssignor, Configurable {
+public class StreamPartitionAssignor implements PartitionAssignor, Configurable, TaskIdToPartitionsProvider {
```

Comment:
I agree that we can leave this in a follow-up PR to keep it minimum. Just leaving more details on my proposal:

1. Remove the `ThreadDataProvider` interface, not let StreamThread to extend it.
2. For all its interface functions, builder / name / processId / config / partitionGrouper can be set in the `configure` call: the thread can set these when constructing the config map for the consumer, which will be passed in the `configure` function; or they can be passed by TaskManager as well (see below)
3. In `StreamThread`, it only needs to call the `activeTasks()` and `standbyTasks()` functions of the assignor. We can move these two functions to the `TaskManager` so that StreamThread would not need to call `setPartitionAssignor` and hold a reference on `assignor` any more, it only needs to call `taskManager.activeTasks()` etc.
4. TaskManager now needs `activeTasks` and `standbyTasks`, and the assignor now needs `prevActiveTasks` and `cachedTasks`. One idea is to extract the task assignment logic from `StreamPartitionAssignor` to `TaskManager`, so that the StreamPartitionAssignor becomes a thin class which only needs to hold a reference of the `PartitionGrouper` and `TaskManager` to call their APIs to construct the map from `partition -> task` and `task -> client-host` to implement the `PartitionAssignor` interface. In this case it does not need `builder, prevActiveTasks, cachedTasks, processId` any more, only `name` for log4j.